pkgdb/src/ld-floxlib.c

```diff
@@ -97,54 +110,110 @@ la_objsearch( const char * name, uintptr_t * cookie, unsigned int flag )
+                  char * flox_env_library_dir = NULL;
+                  char * saveptr              = NULL;  // For strtok_r() context
+
+                  flox_env_library_dir
+                    = strtok_r( flox_env_lib_dirs_buf, ":", &saveptr );
+                  while ( flox_env_library_dir != NULL )
                     {
-                      close( fd );
-                      if ( audit_ld_floxlib < 0 )
+                      if ( flox_env_lib_dirs_count
+                           >= FLOX_ENV_LIB_DIRS_MAXENTRIES )
                         {
-                          audit_ld_floxlib
-                            = ( getenv( "LD_FLOXLIB_AUDIT" ) != NULL );
+                          fprintf( stderr,
+                                   "ERROR: la_objsearch() "
+                                   "FLOX_ENV_LIB_DIRS has too many entries, "
+                                   "truncating to the first %d\n",
+                                   FLOX_ENV_LIB_DIRS_MAXENTRIES );
+                          break;
                         }
-                      if ( audit_ld_floxlib || debug_ld_floxlib )
+                      if ( debug_ld_floxlib )
                         {
                           fprintf( stderr,
-                                   "AUDIT: la_objsearch() resolved %s -> %s\n",
-                                   name,
-                                   name_buf );
+                                   "DEBUG: la_objsearch() "
+                                   "flox_env_lib_dirs[%d] = %s\n",
+                                   flox_env_lib_dirs_count,
+                                   flox_env_library_dir );
                         }
-                      return name_buf;
+                      flox_env_lib_dirs[flox_env_lib_dirs_count]
+                        = flox_env_library_dir;
+                      flox_env_library_dir = strtok_r( NULL, ":", &saveptr );
+                      flox_env_lib_dirs_count++;
                     }
-                  flox_env_library_dir = strtok( NULL, ":" );
                 }
             }
+
+          // Iterate over the list of paths in flox_env_lib_dirs looking for
+          // the requested library. If found, return the full path to the
+          // library and otherwise return the original name.
+          static int i;
+          for ( i = 0; i < flox_env_lib_dirs_count; i++ )
+            {
+              {
+                (void) snprintf( name_buf,
+                                 sizeof( name_buf ),
+                                 "%s/%s",
+                                 flox_env_lib_dirs[i],
+                                 basename );
+                if ( debug_ld_floxlib )
+                  {
+                    fprintf( stderr,
+                             "DEBUG: la_objsearch() checking: %s\n",
+                             name_buf );
+                  }
+                fd = open( name_buf, O_RDONLY );
```

Comment:
Not sure if it matters in this context, but `access(name_buf, F_OK)` would check for existence without checking any permissions.  I could potentially see a difference in behavior where the file existed, but without read permissions.  Failing to _find_ the file here could be confusing, rather than allowing it to be found and report permissions issues later in the load.  I'm unsure of the preferred behavior, but thought I'd mention it.

Reply:
Good question?!? We're operating here in the context of a callback performed by `rtld`, and if we are unable to open the file then I'm not sure there's really any point serving it up as a potential match. Moreover, if we serve up a file that we cannot open then that will stop the search prematurely, potentially before finding an alternative match later in the FLOX_ENV_LIB_DIRS search path. On that basis I'd keep this as an `open()` - @billlevine does that make sense to you?

Reply:
Yep, sure does.